0011-B-Distributed-Lock-building-block.md

```diff
@@ -0,0 +1,359 @@
+
+However, since the current APIs have been present since Dapr 1.8 and haven't changed since they have been introduced, the proposal is to:
+
+- The new APIs will have version "alpha2".
+- Maintain the existing APIs as "alpha1" for one more Dapr release, showing a warning that they are deprecated and will be removed in the next Dapr version.
+- Because the proposed changes impact both components and runtime, this may require keeping the existing code (especially for the existing component) in a "legacy" folder temporarily, for one release.
+
+### Components interface
+
+The interface for all lock components is expanded to the following:
+
+```go
+// Store is the interface for lock stores.
+type Store interface {
+ 	metadata.ComponentWithMetadata
+
+ 	// Init the lock store.
+ 	// Common metadata properties include:
+ 	// - "ttl": default TTL for locks (as a Go duration)
+ 	Init(ctx context.Context, metadata Metadata) error
+
+ 	// Features returns the list of supported features.
+ 	Features() []Feature
+
+ 	// Lock acquires a lock.
+ 	// By default, if the lock is owned by someone else, this method blocks until the lock can be acquired or the context is canceled.
+ 	Lock(ctx context.Context, req LockRequest) (LockResponse, error)
+
+ 	// TryLock tries to acquire a lock.
+ 	// If the lock cannot be acquired, it returns immediately with the error ErrResourceLocked.
+ 	TryLock(ctx context.Context, req LockRequest) (LockResponse, error)
+
+ 	// RenewLock attempts to renew a lock if the lock is still valid.
+ 	// If the lock ID is invalid, or the resource is not locked, returns ErrLockNotFound.
+ 	RenewLock(ctx context.Context, req RenewLockRequest) error
+
+ 	// Unlock tries to release a lock if the lock is still valid.
+ 	// If the lock ID is invalid, or the resource is not locked, returns ErrLockNotFound.
+ 	Unlock(ctx context.Context, req UnlockRequest) error
+}
+
+var (
+	ErrResourceLocked = errors.New("the resource is already locked")
+ 	ErrLockNotFound   = errors.New("the resource is locked with a different lock ID or the lock does not exist")
+)
+
+// LockRequest is the request to acquire locks, used by Lock and TryLock.
+type LockRequest struct {
+	// ID of the resource to lock.
+ 	// This is an arbitrary value supplied by the user.
+ 	ResourceID string `json:"resourceId"`
+ 	// Maximum duration for the lock (if not renewed).
+ 	// If <= 0, the lock uses the default TTL set in the component.
+ 	TTL time.Duration `json:"ttl"`
+}
+
+// LockResponse is the response used by Lock and TryLock when the operation is completed.
+type LockResponse struct {
+ 	// ID of the lock.
+ 	LockID  string `json:"lockID"`
+}
+
+// RenewLockRequest is a lock renewal request.
+type RenewLockRequest struct {
+ 	// ID of the resource whose lock needs to be renewed.
+ 	// This is an arbitrary value supplied by the user.
+ 	ResourceID string `json:"resourceId"`
```

Comment:
We are a bit inconsistent in the Dapr project about `Id` vs `ID` (even in this doc), and even between variable names and their json field encodings. I think we should try and be consistent moving forward.. I am firmly in the `ID` camp :slightly_smiling_face: 
FWIW, Kubernetes also goes with `ID`.